src/v/iceberg/snapshot_json.cc

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/snapshot_json.h"
+
+#include "container/chunked_hash_map.h"
+#include "iceberg/json_utils.h"
+#include "iceberg/snapshot.h"
+#include "json/document.h"
+#include "model/timestamp.h"
+#include "strings/string_switch.h"
+
+namespace iceberg {
+
+namespace {
+
+constexpr std::string_view operation_to_str(snapshot_operation o) {
+    switch (o) {
+    case snapshot_operation::append:
+        return "append";
+    case snapshot_operation::replace:
+        return "replace";
+    case snapshot_operation::overwrite:
+        return "overwrite";
+    case snapshot_operation::delete_data:
+        return "delete";
+    }
+}
+
+snapshot_operation operation_from_str(const ss::sstring& operation_str) {
+    return string_switch<snapshot_operation>(operation_str)
+      .match(
+        operation_to_str(snapshot_operation::append),
+        snapshot_operation::append)
+      .match(
+        operation_to_str(snapshot_operation::replace),
+        snapshot_operation::replace)
+      .match(
+        operation_to_str(snapshot_operation::overwrite),
+        snapshot_operation::overwrite)
+      .match(
+        operation_to_str(snapshot_operation::delete_data),
+        snapshot_operation::delete_data);
+}
+
+} // namespace
+
+snapshot parse_snapshot(const json::Value& v) {
+    auto id = parse_required_i64(v, "snapshot-id");
+    std::optional<snapshot_id> parent_id;
+    auto parent_id_opt = parse_optional_i64(v, "parent-snapshot-id");
+    if (parent_id_opt.has_value()) {
+        parent_id.emplace(parent_id_opt.value());
+    }
+    auto timestamp_ms = parse_required_i64(v, "timestamp-ms");
+    auto manifest_list_path = parse_required_str(v, "manifest-list");
+    auto schema_id_opt = parse_optional_i32(v, "schema-id");
+    std::optional<schema::id_t> schema_id;
+    if (schema_id_opt.has_value()) {
+        schema_id.emplace(schema_id_opt.value());
+    }
+    const auto& summary_json = parse_required(v, "summary");
+    if (!summary_json.IsObject()) {
+        throw std::invalid_argument(fmt::format(
+          "Expected JSON object to parse field 'summary': {}",
```

Comment:
nit for this and the next `GetType` based exception, when the error is thrown it might be a bit hard to understand, eg: `Expected JSON object to parse field 'summary': string`

Could this be more descriptive like `Expected JSON object to parse field 'summary', found: {}`

Reply:
Done